### ISSUE TITLE
Implement multiple resource queues

### ISSUE THREAD

**Issue by scshunt** [OpenRA/OpenRA#9872]
Currently, there are only two types of resources: "cash" and "resources". This should not be hardcoded and works badly with TS (which has veins), RA2 (which doesn't have resources), and AoE (which has several different primary types of resources), to name a few.

The obvious part of this work is to implement separate traits for each different type of resource, replacing `PlayerResources`, which may have different properties (to begin with, some resources must have capacities; some must not).

The other part of this is to implement RA/TD/D2K's behaviour of using resources to spend for cash. This should be implemented using a separate trait to indicate the two. There are two different proposed designs (note that the naming is tentative in both designs):

The first is to have an `ResourceAlias` trait which would implement the same abstract interface as a regular resource, but be implemented differently:

    ResourceAlias: ore-or-cash
        Aliases: ore, cash

This would simply be a resource which would link back into the others. Spending this resource would go in order of the `Aliases` array, so in this case, ore first, then cash.

The second is to have a `SubstituteResource` trait which would allow spending one resource in place of another:

    SubstituteResource:
        Use: ore
        As: cash
        Priority: 1

Resources would be used in decreasing order of priority, so a positive priority indicates that ore would be used first; a negative priority would be used to spend ore only after cash was exhausted.

Feedback on which design to use is appreciated.

**Comment by pchote** [comment]
The main advantage to the `ResourceAlias` is that it comes for free - no additional plumbing or coupling between traits is required.  The `SubstituteResource` will need an explicit check from either the resource being queried or the things trying to use it.

**Comment by scshunt** [comment]
The more I think about it, the more I think that `ResourceAlias` is the more flexible design, since it provides more control to the client of the resource, and it allows for lots of possible permutations, possibly at the same time. On the flip side, it means that every client needs to be aware of all the possible aliases, and I doubt that flexibility would be needed in actual, reasonable module designs.

**Comment by atlimit8** [comment]
I agree with the preference for `ResourceAlias` concept with it implementing `IResourceType`.
Except I would name it differently:
```yaml
ResourceConversion@Funds:
    Name: Funds
    From: Ore, Gems, Cash
```

**Comment by GraionDilach** [comment]
Could someone elaborate to me why allowing to have multiple `PlayerResources` ain't a way to go?

**Comment by pchote** [comment]
@GraionDilach: It is.  It has just been renamed.

**Comment by Phrohdoh** [comment]
```
[13:09:56]	<Phrohdoh> hi all
[13:11:22]	<Phrohdoh> wrt resources it is great that we're filling better possibilities, but I do feel that I need to say _please_ don't become unnecessarily complex. I'm not saying "dumb it down" but why do we *need* the ratios? etc. Also, if this is the way we're going, I think we should require plenty of documentation (and examples).
[13:13:54]	<pchote> hi
[13:13:56]	<pchote> ratios?
[13:15:32]	<Phrohdoh> something atlimit8 suggested on the issue
[13:15:33]	<pchote> ah, I missed that comment
[13:15:36]	<pchote> I agree with you
[13:17:27]	<Phrohdoh> I don't see why we even need conversions. Explicitly require a resource type and be done with that side. On the income side, make certain incomes directly related to certain resources (ie types, again).
[13:17:54]	<pchote> Phrohdoh: the problem is that the original games have two different types of money that can be spent in the same way
[13:18:12]	<Phrohdoh> I understand this, but why don't we implement both of those as cash?
[13:18:14]	<pchote> we need a way of specifying that units can be bought using cash *or* ore
[13:18:20]	<Phrohdoh> We don't store them in silos independantly.
[13:18:25]	<pchote> we do actually
[13:18:33]	<pchote> silos only apply to ore/tib
[13:18:37]	<pchote> so is capped
[13:18:38]	<Phrohdoh> We do globally, then we fill all silos visually.
[13:18:39]	<pchote> cash is not capped
[13:18:45]	<Phrohdoh> Okay, true.
[13:19:45]	<Phrohdoh> I don't think modders will want to think about that (and the complexity), is all I'm trying to say. Unless they're creating an AoE they want things to be bought with "money", whatever you harvest.
[13:20:00]	<Phrohdoh> So cash-or-(ore|tib) seems like the only sensible thing for us to do.
[13:20:12]	<pchote> yes
[13:20:29]	<pchote> I believe that my suggestion there is the simplest thing that will work
[13:20:40]	<pchote> and that we need a strong justification for doing anything more complicated than that
[13:21:26]	<Phrohdoh> I can agree with that, but not the naming coppro suggested.
[13:22:05]	<pchote> naming is a bit of an issue, yes, because of the confusion with our harvestable resource definitions
[13:22:16]	<pchote> it may be possible to combine those though
[13:22:56]	<pchote> but that would bring its own uglyness
[13:23:48]	<Phrohdoh> Certainly. Let's not take any of this without plenty of descs. 
```

**Comment by GraionDilach** [comment]
I don't see the need of `ResourceAlias` justified neither. A `StoredFirst/ProcessedFirst` boolean is enough per `PlayerResource`. The name property of said `PlayerResource` then can be shared with all resources then, so ore, gems and OilDerricks could reference, say the `Money` PR which has a `ProcessedFirst: false` or `StoredFirst:  true` property for bleed behaviour.

**Comment by pchote** [comment]
@GraionDilach: the problem isn't with storage, but with spending.  Mods need the ability to define that a unit can be bought using either cash (not capped, not harvested) *or* ore/tib, and other mods need the ability to define that a unit requires say stone *and* gold *and* steel.

**Comment by GraionDilach** [comment]
Units could have more `Valued` traits, one per `PlayerResource`.

**Comment by scshunt** [comment]
Not all resources have a stored/processed distinction. The intent is to separate them so that we don't maintain that distinction in C# any more but have capacity configurable on a per-resource basis. One could easily imagine a resource type that doesn't have processed resources, but all resources of that type contribute to capacity. We would then need to indicate somewhere both the type of resources that, say, selling a building gives you and also whether it gives processed or stored resources. Splitting these into two is much cleaner.

**Comment by GraionDilach** [comment]
I'm all up with that. But let's see why I still think your way is fishy.

Currently there are 5 ways to gain income in the released mods.
> Through the `OreRefinery` trait
> Through selling an actor
> Through a crate
> Through the `CashTrickler` trait (Oil Derricks, both in RA and TD)
> Through Supply Trucks (RA only)

Right now, the `PlayerResources` trait handles both the storage management (albeit limited since it's hardcoded to distribute this between all actors with the `Storage` trait) and the differentiation between storage-required (all income involving the `OreRefinery` trait) and storage-ignoring (the other three ways) resource (where it prioritizes spending the resource which isn't stored first).

The main limits are the things I remarked.

The problem I see with your proposal is that you see the harvestable resource types of currently released mods (Ore/Gems, greenTiberium/blueTiberium) as different resources, while I see them as subtypes of the same resource, since both use the same `OreRefinery` trait. Shifting the untangling of these resource subtypes from `PlayerResources` level to `Valued` trait level (as @Phrohdoh shown it on the other day) will also imply that any other way excluding the depositable resource types will involve hacks through their traits/actions to duplicate a behaviour which exists already in the current `PlayerResources` trait and it sounds simpler to me to expand that trait with maintaining the resource aliases, unhardcoding the storage usage behaviour (which involves killing off storage for that particular resource as well) and the storaged-notstoraged resource-usage priority (which ofcourse would only apply if the actual resource type could be stored afterall).

I'm fine with using the Valued trait for untangling distinct resource types (dictionary vs multiple traits is just a minor difference) but using it even for untangling resource subtypes is something I find an overhead considering the resources of the currently released mods.

Yes, I'm aware that most games involving multiple resources does not use a storage logic, but that alone does not enough to me to justify an entire trait-interaction logic refactor into something which ultimately scales well with individual resources, but gets messy with interchangeable resources when the default mods have the latter - and only TS will have noninterchangeable resources in the form of cash vs. veins.

**Comment by pchote** [comment]
@GraionDilach: I don't think anyone is proposing that we track the different harvestable tiberium types as separate spendable resources.  They would all be stored and spent as a single Tiberium type.

**Comment by x-a-n-a-x** [comment]
@GraionDilach: Also, (like in Total Annihilation):
- Through salvaged husks or wreckages

**Comment by KOYK** [comment]
is any one working on this?

**Comment by KOYK** [comment]
I can't express enough how much this is needed for other mods.
